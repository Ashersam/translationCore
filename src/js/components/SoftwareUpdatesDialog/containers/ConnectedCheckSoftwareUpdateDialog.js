import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import { getTranslate } from '../../../selectors';
import appPackage from '../../../../../package';
import axios from 'axios';
import os from 'os';
import semver from 'semver';
import CheckSoftwareUpdatesDialog, {
  STATUS_ERROR, STATUS_OK, STATUS_LOADING, STATUS_UPDATE
} from '../components/CheckSoftwareUpdateDialog';

/**
 * Returns the correct update asset for this operating system.
 * If the update is not newer than the installed version null will be returned.
 * @param {object} response the network response
 * @param {string} installedVersion the installed version of the application
 * @param {string} osArch the operating system architecture
 * @param {string} osPlatform the operating system.
 * @return {*} the update object
 */
export function getUpdateAsset(response, installedVersion, osArch, osPlatform) {
  const platformNames = {
    'aix': 'linux',
    'darwin': 'macos',
    'freebsd': 'linux',
    'linux': 'linux',
    'openbsd': 'linux',
    'sunos': 'linux',
    'win32': 'win'
  };
  const platform = `${platformNames[osPlatform]}-${osArch}`;
  let update = null;
  for (const asset of response.assets) {
    if (true || asset.label.includes(platform)) {
      update = {
        ...asset,
        latest_version: response.tag_name,
        installed_version: installedVersion
      };
      break;
    }
  }

  // validate version
  if(update) {
    const latest = semver.valid(semver.coerce(update.latest_version));
    const installed = semver.valid(semver.coerce(update.installed_version));
    if (!semver.gt(latest, installed)) update = null;
  }
  return update;
}

export class ConnectedSoftwareUpdateDialog extends React.Component {

  constructor(props) {
    super(props);
    this.handleClose = this.handleClose.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.startSoftwareCheck = this.startSoftwareCheck.bind(this);
    this.stopSoftwareCheck = this.stopSoftwareCheck.bind(this);
    this.initialState = {
      status: STATUS_LOADING,
      update: null,
      cancelToken: null
    };
    this.state = {
      ...this.initialState
    };
  }

  componentWillReceiveProps(newProps) {
    const openChanged = newProps.open !== this.props.open;
    if(openChanged && newProps.open) {
      this.startSoftwareCheck();
    } else if(openChanged && !newProps.open) {
      this.stopSoftwareCheck();
    }
  }

  componentDidCatch(error, info) {
    console.error(error, info);
    this.stopSoftwareCheck();
  }

  componentWillUnmount() {
    this.stopSoftwareCheck();
  }

  /**
   * Initiates checking for software updates
   */
  startSoftwareCheck() {
    const CancelToken = axios.CancelToken;
    const source = CancelToken.source();
    const request = {
      cancelToken: source.token,
      method: 'GET',
      url: `https://api.github.com/repos/unfoldingWord-dev/translationCore/releases/latest`
    };

    this.setState({
      ...this.initialState,
      cancelToken: source
    });

    this.request = axios(request).then(response => {
      const update = getUpdateAsset(response.data, appPackage.version, os.arch(), os.platform());
      if(update) {
        this.setState({
          ...this.state,
          status: STATUS_UPDATE,
          update
        });
      } else {
        this.setState({
          ...this.state,
          status: STATUS_OK
        });
      }
    }).catch(error => {
      if(axios.isCancel(error)) {
        // user canceled
        this.handleClose();
      } else {
        console.error(error);
        this.setState({
          ...this.state,
          status: STATUS_ERROR
        });
      }
    });
  }

  /**
   * Cancels the software update checks
   */
  stopSoftwareCheck() {
    const {cancelToken} = this.state;
    if(cancelToken !== null) {
      cancelToken.cancel('Operation canceled by user');
    }
  }

  handleClose() {
    const {onClose} = this.props;
    this.setState({
      ...this.initialState
    });
    onClose();
  }

  handleSubmit() {
    const {update} = this.state;
    console.log(update);
    const {onDownload} = this.props;
    if(update) {
      onDownload({
        content_type: update.content_type,
        name: update.name,
        version: update.latest_version,
        size: update.size,
        url: update.browser_download_url
      });
    } else {
      this.handleClose();
    }
  }

  render() {
    const {open, translate} = this.props;
    const {status, update} = this.state;

    return (
      <CheckSoftwareUpdatesDialog onClose={this.handleClose}
                                  translate={translate}
                                  open={open}
                                  onSubmit={this.handleSubmit}
                                  status={status}
                                  update={update}/>
    );
  }
}

ConnectedSoftwareUpdateDialog.propTypes = {
  translate: PropTypes.func,
  onClose: PropTypes.func.isRequired,
  open: PropTypes.bool.isRequired,
  onDownload: PropTypes.func.isRequired
};

const mapStateToProps = (state) => ({
  translate: getTranslate(state)
});

export default connect(mapStateToProps)(ConnectedSoftwareUpdateDialog);
