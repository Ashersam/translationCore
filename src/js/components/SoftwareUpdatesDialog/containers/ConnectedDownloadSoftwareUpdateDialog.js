import React from 'react';
import PropTypes from 'prop-types';
import DownloadDialog from '../components/DownloadDialog';
import axios from 'axios';
import { getTranslate } from '../../../selectors';
import {connect} from 'react-redux';
import { ipcRenderer } from 'electron';
import fs from 'fs-extra';

/**
 * @deprecated remove this from package.json
 */
import fileDownload from 'react-file-download';

/**
 * Renders a dialog to download the software update.
 * This component will begin downloading as soon as it is mounted.
 */
class ConnectedDownloadSoftwareUpdateDialog extends React.Component {

  constructor(props) {
    super(props);
    this.handleClose = this.handleClose.bind(this);
    this.cancelDownload = this.cancelDownload.bind(this);
    this.initialState = {
      downloaded: 0,
      indeterminate: true,
      total: props.update.size,
      cancelToken: null,
      error: false
    };
    this.state = {
      ...this.initialState
    };
  }

  componentDidCatch(error, info) {
    const {onClose} = this.props;
    console.error(error, info);
    onClose();
  }

  componentWillUnmount() {
    this.cancelDownload();
  }

  componentDidMount() {
    const {update} = this.props;

    ipcRenderer.send('download-as', {
      filename: update.name,
      url: update.url
    });

    // TODO: retrieve the download item so we can cancel it

    ipcRenderer.on('download-as-success', (event, args) => {
      console.log('success', args);
      this.handleClose();
    });
    ipcRenderer.on('download-as-progress', (event, progress) => {
      this.setState({
        ...this.state,
        indeterminate: false,
        headers: {
          Accept: update.content_type
        },
        downloaded: update.size * progress
      });
    });
    ipcRenderer.on('download-as-error', (event, args) => {
      console.log(args);
      this.setState({
        ...this.initialState,
        error: true
      });
    });
    this.setState({
      ...this.initialState,
      indeterminate: true
    });
    return;

    const source = axios.CancelToken.source();
    const request = {
      url: update.url,
      method: 'get',
      responseType: 'arraybuffer',
      onDownloadProgress: event => {
        this.setState({
          ...this.state,
          indeterminate: false,
          headers: {
            Accept: update.content_type
          },
          downloaded: event.loaded,
          total: event.total
        });
      },
      cancelToken: source.token
    };

    this.setState({
      ...this.initialState,
      cancelToken: source,
      indeterminate: true
    });

    axios(request).then(response => {
      const filePath = ipcRenderer.sendSync('save-as', { options: {
          defaultPath: update.name
      }});
      if(!filePath) {
        this.handleClose();
      } else {
        // TODO: save the file
        fs.outputFileSync(filePath, response.data);
      }
      // return fileDownload(response.data, update.name, update.content_type);
    }).then(() => {
      this.handleClose();
    }).catch(error => {
      if(!axios.isCancel(error)) {
        console.error('Failed to download app update', error);
        this.setState({
          ...this.initialState,
          error: true
        });
      }
    });
  }

  cancelDownload() {
    const {cancelToken} = this.state;
    if(cancelToken) {
      cancelToken.cancel('Operation canceled by user');
    }
  }

  handleClose() {
    const {onClose} = this.props;
    this.cancelDownload();
    onClose();
  }

  render() {
    const {translate, update, open} = this.props;
    const {downloaded, indeterminate, total, error} = this.state;
    let message = (
      <div>
        <p>
          {translate('software_update.downloading_version', {
            version: update.version
          })}
        </p>
        <p>
          {Math.round(downloaded/1024/1025)}/{Math.round(total/1024/1025)} MB
        </p>
      </div>
    );
    const title = translate('software_update.downloading');
    if(error) {
      message = (
        <p>
          {translate('software_update.download_error')}
        </p>
      );
    }
    return <DownloadDialog message={message}
                           size={total}
                           open={open}
                           indeterminate={indeterminate}
                           sizeDownloaded={downloaded}
                           cancelLabel={translate('cancel')}
                           onCancel={this.handleClose}
                           title={title}/>;
  }
}

ConnectedDownloadSoftwareUpdateDialog.propTypes = {
  open: PropTypes.bool.isRequired,
  translate: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired,
  update: PropTypes.object.isRequired
};
const mapStateToProps = (state) => ({
  translate: getTranslate(state)
});
export default connect(mapStateToProps)(ConnectedDownloadSoftwareUpdateDialog);
