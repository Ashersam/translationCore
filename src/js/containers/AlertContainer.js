import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { getAlerts } from '../selectors';
import { closeAlert, ignoreAlert } from '../actions/AlertActions';
import Alert from './dialogs/Alert';

class AlertPortal extends React.PureComponent {
  constructor (props) {
    super(props);
    this.el = document.createElement('div');
  }

  componentDidMount () {
    document.body.appendChild(this.el);
  }

  componentWillUnmount () {
    document.body.removeChild(this.el);
  }

  render () {
    return ReactDOM.createPortal(this.props.children, this.el);
  }
}

AlertPortal.propTypes = {
  children: PropTypes.any.isRequired
};

class Alerts extends React.Component {

  constructor(props) {
    super(props);
    this.handleOnIgnore = this.handleOnIgnore.bind(this);
    this.handleOnCancel = this.handleOnCancel.bind(this);
    this.handleOnConfirm = this.handleOnConfirm.bind(this);
  }

  handleOnCancel(alert) {
    const {closeAlert} = this.props;
    if(typeof alert.onCancel === 'function') {
      return () => {
        closeAlert(alert.id);
        // propagate callback
        alert.onCancel();
      };
    }
  }

  handleOnConfirm(alert) {
    const {closeAlert} = this.props;
    return () => {
      closeAlert();
      // propagate callback
      if(typeof alert.onConfirm === 'function') {
        alert.onConfirm();
      }
    };
  }

  handleOnIgnore(alert) {
    const {ignoreAlert} = this.props;
    return ignored => {
      ignoreAlert(alert.id, ignored);
      // propagate callback
      if(typeof alert.onIgnore === 'function') {
        alert.onIgnore(ignored);
      }
    };
  }

  render () {
    const alerts = this.props.alerts.map((item, i) => (
      <AlertPortal key={i}>
        <Alert {...item}
               onIgnore={this.handleOnIgnore(item)}
               onCancel={this.handleOnCancel(item)}
               onConfirm={this.handleOnConfirm(item)}
               open={true}/>
      </AlertPortal>
    ));

    return (
      <div className="modals">
        {alerts}
      </div>
    );
  }
}

Alerts.propTypes = {
  alerts: PropTypes.array,
  closeAlert: PropTypes.func.isRequired,
  ignoreAlert: PropTypes.func.isRequired
};
Alerts.defaultProps = {
  alerts: []
};

const mapStateToProps = (state) => {
  return {
    alerts: getAlerts(state)
  };
};
const mapDispatchToProps = {
  closeAlert,
  ignoreAlert
};

export default connect(mapStateToProps, mapDispatchToProps)(Alerts);
