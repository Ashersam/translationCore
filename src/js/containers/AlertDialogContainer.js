import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withTranslate } from '../components/Locale';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
// components
import Alert from '../components/dialogComponents/Alert';
// actions
import { closeAlertDialog } from '../actions/AlertModalActions';

class AlertDialogContainer extends Component {

  render(){
    const {translate} = this.props;
    return (
      <div>
        <MuiThemeProvider>
          <Alert translate={translate}
                 {...this.props}/>
        </MuiThemeProvider>
      </div>
    );
  }
}

AlertDialogContainer.propTypes = {
  translate: PropTypes.func.isRequired
};

const mapStateToProps = (state) => {
  return {
    alertModalReducer: state.alertModalReducer,
    translate: withTranslate(state)
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    actions: {
      closeAlertDialog: () => {
        dispatch(closeAlertDialog());
      }
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AlertDialogContainer);
