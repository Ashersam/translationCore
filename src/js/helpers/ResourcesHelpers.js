
import fs from 'fs-extra';
import path from 'path-extra';
// constant declarations
const BIBLE_RESOURCES_PATH = path.join(path.homedir(), 'translationCore/resources/bibles');
const THELPS_RESOURCES_PATH = path.join(path.homedir(), 'translationCore/resources/translationHelps');
const STATIC_RESOURCES_BIBLES_PATH = './static/resources/bibles';
const STATIC_RESOURCES_THELPS_PATH = './static/resources/translationHelps';

/**
 * @description moves all bibles from the static folder to the local user translationCore folder.
 */
export function getBibleFromStaticPackage() {
  let bibleNames = fs.readdirSync(STATIC_RESOURCES_BIBLES_PATH);
  bibleNames.forEach((bibleName) => {
    let bibleSourcePath = path.join(STATIC_RESOURCES_BIBLES_PATH, bibleName);
    let bibleDestinationPath = path.join(BIBLE_RESOURCES_PATH, bibleName);
    if(!fs.existsSync(bibleDestinationPath)) {
      fs.copySync(bibleSourcePath, bibleDestinationPath);
    }
  });
}

/**
 * @description moves all translationHelps from the static folder to the resources folder in the translationCore folder.
 */
export function getTHelpsFromStaticPackage() {
  let tHelpsNames = fs.readdirSync(STATIC_RESOURCES_THELPS_PATH);
  tHelpsNames.forEach((tHelpName) => {
    let tHelpSourcePath = path.join(STATIC_RESOURCES_THELPS_PATH, tHelpName);
    let tHelpDestinationPath = path.join(THELPS_RESOURCES_PATH, tHelpName);
    if(!fs.existsSync(tHelpDestinationPath)) {
      fs.copySync(tHelpSourcePath, tHelpDestinationPath);
    }
  });
}