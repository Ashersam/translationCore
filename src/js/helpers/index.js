import {generateTimestamp} from './TimestampGenerator.js';

export {
  generateTimestamp
};
