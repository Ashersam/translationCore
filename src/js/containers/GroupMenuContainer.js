/* eslint-disable react/no-find-dom-node */
import React from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux';
import { Grid, Col, Glyphicon } from 'react-bootstrap';
import * as style from '../components/groupMenu/Style';
// components
import GroupsMenuFilter from '../components/groupMenu/GroupsMenuFilter';
import Groups from '../components/groupMenu/Groups';
import Group from '../components/groupMenu/Group';
import GroupItem from '../components/groupMenu/GroupItem';
// actions
import { changeCurrentContextId } from '../actions/ContextIdActions.js';
import { expandSubMenu } from '../actions/GroupMenuActions.js';
import * as CheckDataLoadActions from '../actions/CheckDataLoadActions';
//helpers
import * as ProjectDetailsHelpers from '../helpers/ProjectDetailsHelpers';
import isEqual from 'deep-equal';
import * as statusBadgeHelpers from '../helpers/statusBadgeHelpers';

const MENU_BAR_HEIGHT = 30;
const MENU_ITEM_HEIGHT = 38;

const groupMenuContainerStyle = {
  backgroundColor: "var(--background-color-dark)",
  zIndex: "98",
  fontSize: "12px",
  overflowX: "hidden",
  height: "100%",
  padding: 0,
  position: "fixed",
  width: "250px"
};

// Use named export for unconnected component (for tests)
export class GroupMenuContainer extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      showInvalidated: false,
      showBookmarks: false,
      showSelections: false,
      showNoSelections: false,
      showVerseEdits: false,
      showComments: false,
      expandFilter: false
    };
  }

  setFilter(name, value) {
    this.setState({[name]: value});
    if (name == 'showSelections' && value) {
      this.setState({showNoSelections: false});
    }
    if (name == 'showNoSelections' && value) {
      this.setState({showSelections: false});
    }
  }

  handleFilterToggle() {
    this.setState({expandFilter: !this.state.expandFilter});
  }

  menu() {
    const { translate, toolsReducer: {currentToolName} } = this.props;
    let menu = <div />;
    if (currentToolName !== null) {
      const filterCount = this.countFilters();      
      menu = (
        <div id="groups-menu-container">
          <div id="groups-menu-top">
            <div id="groups-menu-header">
              <span id="groups-menu-title">
                {this.props.translate('tools.menu')}
              </span>
              { currentToolName==="translationWords" ?
                <Glyphicon
                  key="filter"
                  glyph="filter"
                  className={'filter-icon '+(this.state.expandFilter?'expanded':'collapsed')}
                  onClick={this.handleFilterToggle.bind(this)} />
              :''}
              {!this.state.expandFilter && filterCount?<span className="filter-badge badge" onClick={this.handleFilterToggle.bind(this)}>{filterCount}</span>:""}
              </div>
            {currentToolName==="translationWords"?
              <GroupsMenuFilter
                {...this.state}
                currentToolName={currentToolName}
                translate={translate}
                setFilter={this.setFilter.bind(this)} />
              :''}
          </div>
          <Groups groups={this.groups()} />
        </div>
      );
    }
    return menu;
  }

  /**
  * @description - Tests if the the two elements are in the scope of the window (scroll bar)
  * The consts MENU_BAR_HEIGHT & MENU_ITEM_HEIGHT are set to account for the static window avialablity
  * @param {object} groupMenu - The current group menu header that is extended/actived (i.e. Metaphors)
  * @param {object} currentItem - The current group check item that is active (i.e. Luke 1:1)
  */
  inView(groupMenu, currentItem) {
    var rectGroup = ReactDOM.findDOMNode(groupMenu).getBoundingClientRect();
    var rectItem = ReactDOM.findDOMNode(currentItem).getBoundingClientRect();
    var viewHeight = Math.max(document.documentElement.clientHeight, window.innerHeight);
    return Math.abs(rectGroup.top - rectItem.top) + MENU_BAR_HEIGHT + MENU_ITEM_HEIGHT <= viewHeight;
  }

  getGroupData(groupsData, groupId) {
    let groupData;
    if (groupsData !== undefined) {
      groupData = groupsData[groupId];
    }
    return groupData;
  }

  /**
  * @description generates the total progress for the group.
  * @return {Number} - progress percentage.
  */
  generateProgress(groupIndex) {
    let { groupsData } = this.props.groupsDataReducer;
    let groupId = groupIndex.id;
    let totalChecks = groupsData[groupId].length;
    let doneChecks = 0;

    groupsData[groupId].forEach(function (groupData) {
      if (groupData.selections && !groupData.reminders) {
        doneChecks++;
      }
    }, this);

    let progress = doneChecks / totalChecks;

    return progress;
  }

  /**
  * @description gets the group data for the groupItem.
  * @return {object} groud data object.
  */
  getItemGroupData(contextId, groupIndex) {
    let { groupsData } = this.props.groupsDataReducer;
    let groupId = groupIndex.id;

    let groupData = groupsData[groupId].filter(groupData => {
      return isEqual(groupData.contextId, contextId);
    });

    return groupData[0];
  }

  /**
   * @description - gets the status badge component for the group menu row
   * @param {object} groupItemData 
   */
  getStatusBadge(groupItemData) {
    if (!groupItemData || ! groupItemData.contextId)
    console.log(groupItemData);
    const { chapter, verse } = groupItemData.contextId.reference;
    const { alignmentData } = this.props.wordAlignmentReducer;
    const wordBank = alignmentData && alignmentData[chapter] && alignmentData[chapter][verse] ? alignmentData[chapter][verse].wordBank : [];
    const { currentToolName } = this.props.toolsReducer;
    const glyphs = [];

    // The below ifs are in order of precedence of the status badges we show
    // TODO: groupItemData should have an `invalidated` boolean when invalidation is done for all verses in #3086
    if (groupItemData.invalidated) glyphs.push('invalidated');
    if (groupItemData.reminders)   glyphs.push('bookmark');
    if (groupItemData.selections || (currentToolName === 'wordAlignment' && wordBank && wordBank.length === 0))
      glyphs.push('ok');
    if (groupItemData.verseEdits)  glyphs.push('pencil');
    if (groupItemData.comments)    glyphs.push('comment');
      
    return statusBadgeHelpers.getStatusBadge(glyphs);
  }

  scrollIntoView(element) {
    ReactDOM.findDOMNode(element).scrollIntoView({ block: 'end', behavior: 'smooth' });
  }

  /**
   * @description - Determines if the item should be shown in the group menu based on filters
   * @param {object} groupItemData 
   * @returns {boolean}
   */
  showGroupItem(groupItemData) {
    const {
      showInvalidated,
      showBookmarks,
      showSelections,
      showNoSelections,
      showVerseEdits,
      showComments
    } = this.state;
    return (! this.countFilters() ||
      ((showInvalidated && groupItemData.invalidated) 
        || (showBookmarks && groupItemData.reminders)
        || (showSelections && groupItemData.selections)
        || (showNoSelections && ! groupItemData.selections)
        || (showVerseEdits && groupItemData.verseEdits)
        || (showComments && groupItemData.comments)
      ));
  }

  /**
   * @description - Determines if the group should be shown in the group menu based on filters
   * @param {object} groupData
   * @returns {boolean}
   */
  showGroup(groupData) {
    if (! this.countFilters()) {
      return true;
    }
    for (let groupItemData of groupData) {
      if (this.showGroupItem(groupItemData)) {
        return true;
      }
    }
    return false;
  }

  countFilters() {
    const { showInvalidated, showBookmarks, showSelections, showNoSelections, showVerseEdits, showComments } = this.state;
    return [showInvalidated, showBookmarks, showSelections, showNoSelections, showVerseEdits, showComments].filter(k=>k).length;
  }

  /**
  * @description Maps all groupData aka check objects to GroupItem components
  * @param {array} groupData - array of all groupData objects
  * @param {object} groupIndex
  * @param {object} groupHeaderComponent
  * @return {array} groupItems - array of groupData mapped to GroupItem components
  */
  getGroupItemComponents(groupData, groupIndex, groupHeaderComponent) {
    const contextIdReducer = {...this.props.contextIdReducer};
    const projectDetailsReducer = {...this.props.projectDetailsReducer};
    const { manifest } = this.props.projectDetailsReducer;
    const items = [];
    let index = 0;
    for (let groupItemData of groupData) {
      if (! this.showGroupItem(groupItemData)) {
        continue;
      }
      contextIdReducer.contextId = groupItemData.contextId;
      let loadPath = CheckDataLoadActions.generateLoadPath(projectDetailsReducer, contextIdReducer, 'selections');
      let selectionsObject = CheckDataLoadActions.loadCheckData(loadPath, groupItemData.contextId);
      let selectionsArray = [];

      if (selectionsObject) {
        selectionsObject.selections.forEach((selection) => {
          selectionsArray.push(selection.text);
        });
      }
      let selections = selectionsArray.join(" ");

      let active = isEqual(groupItemData.contextId, this.props.contextIdReducer.contextId);
      let useTargetLanguageBookName = manifest.target_language && manifest.target_language.book && manifest.target_language.book.name;
      let bookName = useTargetLanguageBookName ? manifest.target_language.book.name : manifest.project.name;

      if (selections) {
        // Convert the book name to the abbreviation tit -> Tit
        let bookAbbr = manifest.project.id;
        bookName = bookAbbr.charAt(0).toUpperCase() + bookAbbr.slice(1);
      }

      items.push(
        <GroupItem
          key={index}
          {...this.props}
          {...groupItemData}
          statusBadge={this.getStatusBadge(groupItemData)}
          groupMenuHeader={groupHeaderComponent}
          scrollIntoView={this.scrollIntoView.bind(this)}
          active={active}
          bookName={bookName}
          selectionText={selections}
          inView={this.inView}
        />
      );
      index++;
    }
    return items;
  }

  /**
  * @description converts groupsIndex into array of Group components
  * @param {array} groupsIndex - array of all groupIndex objects
  * @return {array} groups - array of Group components
  */
  groups() {
    let { groupsIndex } = this.props.groupsIndexReducer;
    let groupComponents = (<div className='no-results'>{this.props.translate('tools.no_results')}</div>);
    let { groupsData } = this.props.groupsDataReducer;
    let { projectSaveLocation } = this.props.projectDetailsReducer;
    let progress;

    if (groupsIndex !== undefined) {
      groupsIndex = groupsIndex.filter(groupIndex => {
        return groupsData !== undefined && Object.keys(groupsData).includes(groupIndex.id) && this.showGroup(this.getGroupData(groupsData, groupIndex.id));
      });
      if (groupsIndex.length) {
        groupComponents = groupsIndex.map(groupIndex => {
          let { contextId } = this.props.contextIdReducer;
          let groupId = groupIndex.id;
          let currentGroupData = this.getGroupData(groupsData, groupId);
          let active = false;
          
          if (contextId !== null) {
            active = contextId.groupId === groupId;
          }
          
          if (contextId && contextId.tool === 'wordAlignment') {
            progress = ProjectDetailsHelpers.getWordAlignmentProgressForGroupIndex(projectSaveLocation, contextId.reference.bookId, groupIndex);
          } else {
            progress = this.generateProgress(groupIndex);
          }
          
          const getGroupItems = (groupHeaderComponent) => {
            return this.getGroupItemComponents(currentGroupData, groupIndex, groupHeaderComponent);
          };
          
          return (
            <Group
              {...this.props}
              getGroupItems={getGroupItems}
              groupIndex={groupIndex}
              active={active}
              key={groupIndex.id}
              progress={progress}
              openGroup={() => this.props.actions.groupMenuChangeGroup(currentGroupData[0].contextId)}
            />
          );
        });
      }
    }
    return groupComponents;
  }

  render() {
    let { onToggleMenu } = this.props.actions;
    let { menuVisibility } = this.props.groupMenuReducer;
    return (
      <div className="group-menu">
        <div style={{ display: menuVisibility ? "block" : "none" }}>
          <Grid fluid style={groupMenuContainerStyle}>
            <Col style={
              {
                width: "250px",
                position: "fixed",
                padding: 0,
                backgroundColor: "var(--background-color-dark)",
                height: "95%",
              }
            }>
              {this.menu()}
            </Col>
          </Grid>
        </div>
        <Glyphicon
          style={menuVisibility ? style.slideButton : style.slideButtonCollapsed}
          glyph={menuVisibility ? 'chevron-left' : 'chevron-right'}
          onClick={onToggleMenu}
        />
      </div>
    );
  }
}

GroupMenuContainer.propTypes = {
  groupsDataReducer: PropTypes.any.isRequired,
  contextIdReducer: PropTypes.any.isRequired,
  projectDetailsReducer: PropTypes.any.isRequired,
  groupsIndexReducer: PropTypes.any.isRequired,
  actions: PropTypes.any.isRequired,
  groupMenuReducer: PropTypes.any.isRequired,
  toolsReducer: PropTypes.any.isRequired,
  wordAlignmentReducer: PropTypes.any.isRequired,
  translate: PropTypes.func.isRequired
};

const mapStateToProps = (state) => {
  return {
    groupsIndexReducer: state.groupsIndexReducer,
    groupsDataReducer: state.groupsDataReducer,
    selectionsReducer: state.selectionsReducer,
    contextIdReducer: state.contextIdReducer,
    resourcesReducer: state.resourcesReducer,
    projectDetailsReducer: state.projectDetailsReducer,
    groupMenuReducer: state.groupMenuReducer,
    toolsReducer: state.toolsReducer,
    remindersReducer: state.remindersReducer,
    wordAlignmentReducer: state.wordAlignmentReducer,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    actions: {
      changeCurrentContextId: contextId => {
        dispatch(changeCurrentContextId(contextId));
      },
      groupMenuChangeGroup: contextId => {
        dispatch(changeCurrentContextId(contextId));
        dispatch(expandSubMenu(true));
      },
      groupMenuExpandSubMenu: isSubMenuExpanded => {
        dispatch(expandSubMenu(isSubMenuExpanded));
      }
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(GroupMenuContainer);
