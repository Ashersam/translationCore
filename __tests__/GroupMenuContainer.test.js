/* eslint-disable no-console */
/* eslint-env jest */
import * as ResourcesHelpers from "../src/js/helpers/ResourcesHelpers";

jest.unmock('fs-extra');
import React from 'react';
import { GroupMenuContainer } from '../src/js/containers/GroupMenuContainer';
import {shallow} from 'enzyme';
import { Grid, Col, Glyphicon } from 'react-bootstrap';

describe('GroupMenuContainer', () => {
  let contextIdReducer ,wordAlignmentReducer, resourcesReducer, toolsReducer, groupsDataReducer, projectDetailsReducer,
    groupsIndexReducer, groupMenuReducer;
  const luke1 = require('./fixtures/project/wordAlignmentData/luke/luk/1.json');
  const lukeIndex = require('./fixtures/project/wordAlignmentData/luke/index.json');

  beforeEach(() => {
    contextIdReducer = {
      "contextId": {
        "groupId": "figs_metaphor",
        "occurrence": 1,
        "quote": "that he put before them",
        "information": "Paul speaks about good deeds as if they were objects that God could place in front of people. AT: \"that God prepared for them to do\" (See: [[:en:ta:vol1:translate:figs_metaphor]]) \n",
        "reference": {
          "bookId": "luk",
          "chapter": 1,
          "verse": 1
        },
        "tool": "TranslationNotesChecker"
      }
    };
    wordAlignmentReducer = {
      alignmentData: {
        "1": luke1
      }
    };
    toolsReducer = {
      currentToolViews: {},
      currentToolName: 'cheese',
      currentToolTitle: null,
      toolsMetadata:[]
    };
    groupsDataReducer = {
      groupsData: {},
      loadedFromFileSystem: false
    };
    projectDetailsReducer = {
      projectSaveLocation: '',
      manifest: {},
      currentProjectToolsProgress: {},
      projectType: null
    };
    groupsIndexReducer = {
      groupsIndex: lukeIndex,
      loadedFromFileSystem: false
    };
    groupMenuReducer = {
      menuVisibility: true,
      isSubMenuExpanded: true
    };
  });

  test('GroupMenuContainer renders Luke 1:1', () => {
    // given
    const chapter = "1";
    const verse = "1";
    contextIdReducer.contextId.reference.chapter = chapter;
    contextIdReducer.contextId.reference.verse = verse;

    // when
    const enzymeWrapper = shallow(
      <GroupMenuContainer
        groupsDataReducer={groupsDataReducer}
        contextIdReducer={contextIdReducer}
        projectDetailsReducer={projectDetailsReducer}
        groupsIndexReducer={groupsIndexReducer}
        actions={{}}
        groupMenuReducer={groupMenuReducer}
        toolsReducer={toolsReducer}
        wordAlignmentReducer={wordAlignmentReducer}
      />
    );

    // then
    const div = enzymeWrapper.find('div');
    const grid = div.find(Grid);
    expect(grid.length).toEqual(1);
  });
});
