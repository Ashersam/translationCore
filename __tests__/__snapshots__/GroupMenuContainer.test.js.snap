// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupMenuContainer GroupMenuContainer renders the status badge for a completed wordAlignment as "ok" glyph 1`] = `
<div
  className="status-badge"
>
  <Glyphicon
    bsClass="glyphicon"
    glyph="ok"
    style={
      Object {
        "color": "var(--completed-color)",
        "display": "initial",
      }
    }
  />
</div>
`;

exports[`GroupMenuContainer GroupMenuContainer renders the status badge for a not completed wordAlignment as an empty glyph 1`] = `
<div
  className="status-badge"
>
  <Glyphicon
    bsClass="glyphicon"
    glyph=""
    style={
      Object {
        "color": "none",
        "display": "initial",
        "paddingLeft": "15px",
      }
    }
  />
</div>
`;

exports[`GroupMenuContainer GroupMenuContainer renders the status badge for all status booleans 1`] = `
<div
  className="status-badge-wrapper"
>
  <div
    className="status-badge"
    data-class="status-tooltip"
    data-delay-hide="100"
    data-effect="float"
    data-html="true"
    data-offset="{'bottom': -5, 'right': 5}"
    data-place="bottom"
    data-tip="<span style=\\"color:var(--reverse-color);display:initial\\" class=\\"glyphicon glyphicon-bookmark\\" data-reactroot=\\"\\"></span><span style=\\"color:var(--completed-color);display:initial\\" class=\\"glyphicon glyphicon-ok\\" data-reactroot=\\"\\"></span><span style=\\"color:var(--reverse-color);display:initial\\" class=\\"glyphicon glyphicon-pencil\\" data-reactroot=\\"\\"></span><span style=\\"color:var(--highlight-color);display:initial\\" class=\\"glyphicon glyphicon-comment\\" data-reactroot=\\"\\"></span>"
  >
    <div
      className="glyphicon glyphicon-invalidated"
    >
      <InvalidatedIcon
        height={16}
        width={16}
      />
    </div>
    <div
      className="badge"
    >
      5
    </div>
  </div>
  <ReactTooltip
    insecure={true}
    resizeHide={true}
    wrapper="div"
  />
</div>
`;
